src/coreclr/jit/valuenum.cpp

```diff
@@ -4594,6 +4594,88 @@ bool ValueNumStore::IsVNConstant(ValueNum vn)
     }
 }
 
+bool ValueNumStore::IsVNVectorZero(ValueNum vn)
+{
+    if (vn == NoVN)
+    {
+        return false;
+    }
+#if FEATURE_HW_INTRINSICS
+    VNFuncApp funcApp;
+    if (GetVNFunc(vn, &funcApp) && funcApp.m_arity == 1)
+    {
+        switch (funcApp.m_func)
+        {
+            case VNF_HWI_Vector128_get_Zero:
+#if defined(TARGET_XARCH)
+            case VNF_HWI_Vector256_get_Zero:
+#elif defined(TARGET_ARM64)
+            case VNF_HWI_Vector64_get_Zero:
+#endif
+            {
+                return true;
+            }
+
+            default:
+                return false;
+        }
+    }
+#endif
+    return false;
+}
+
+#if FEATURE_SIMD
+VNSimdTypeInfo ValueNumStore::GetSimdTypeOfVN(ValueNum vn)
+{
+    VNSimdTypeInfo vnInfo;
+
+    if (vn == NoVN)
+    {
+        vnInfo.m_simdSize        = 0;
+        vnInfo.m_simdBaseJitType = CORINFO_TYPE_UNDEF;
+        return vnInfo;
+    }
+
+    // The SIMD type is encoded as a function,
+    // even though it is not actually a function.
+    VNFuncApp simdType;
+    if (GetVNFunc(vn, &simdType) && simdType.m_func == VNF_SimdType && simdType.m_arity == 2)
```

Comment:
Arity of a `VNFunc` is fixed (there is one exception that's not relevant here) so there is no need to check it after we've checked the `m_func` (we could assert though).

Reply:
I'm used to checking the 'shape' of information like this rather than being optimistic and putting in asserts; but I think it probably doesn't matter too much TBH. I can put in the assert as that's the philosophy.

Reply:
> I'm used to checking the 'shape' of information like this rather than being optimistic 

Arity of `VNFunc`s being fixed is a pretty strong invariant in VN that will never change, so I think it is absolutely fine to rely on it.